src/main/java/io/quarkus/search/app/quarkiverseio/QuarkiverseIO.java

```diff
@@ -0,0 +1,152 @@
+
+import java.io.Closeable;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+import io.quarkus.search.app.entity.Guide;
+import io.quarkus.search.app.hibernate.InputProvider;
+import io.quarkus.search.app.indexing.reporting.FailureCollector;
+import io.quarkus.search.app.util.CloseableDirectory;
+
+import io.quarkus.logging.Log;
+
+import org.hibernate.search.util.common.impl.Closer;
+
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+
+public class QuarkiverseIO implements Closeable {
+
+    public static final String QUARKIVERSE_ORIGIN = "quarkiverse-hub";
+
+    private final FailureCollector failureCollector;
+
+    private final boolean enabled;
+    private final Path pages;
+    private final URI baseUri;
+    private final CloseableDirectory tempDir;
+
+    public QuarkiverseIO(boolean enabled, Path pages, URI baseUri, FailureCollector failureCollector,
+            CloseableDirectory tempDir) {
+        this.failureCollector = failureCollector;
+        this.enabled = enabled;
+        this.pages = pages;
+        this.baseUri = baseUri;
+        this.tempDir = tempDir;
+    }
+
+    private Guide readGuide(Path file) {
+        Guide guide = new Guide();
+        guide.url = baseUri.resolve(pages.relativize(file).toString());
+        guide.type = "reference";
+        guide.origin = QUARKIVERSE_ORIGIN;
+
+        try {
+            Document document = Jsoup.parse(file);
+
+            String title = document.select("h1.page").text();
+            if (title.isBlank()) {
+                title = document.select("nav.breadcrumbs").text();
+            }
+            if (title.isBlank()) {
+                title = document.select("h3.title").text();
+            }
+            guide.title.set(title.trim());
+
+            guide.summary.set(document.select("div#preamble").text());
+        } catch (IOException e) {
+            failureCollector.warning(FailureCollector.Stage.PARSING, "Failed to parse guide file: " + file, e);
+        }
+
+        guide.htmlFullContentProvider.set(new FileInputProvider(file));
+
+        Log.debugf("Parsed guide: %s", guide.url);
+        return guide;
+    }
+
+    public Stream<Guide> guides() {
+        if (enabled) {
+            try {
+                return Files.list(pages).filter(Files::isDirectory)
+                        // First get the extensions directories:
+                        .filter(dir -> dir.getFileName().toString().startsWith("quarkus"))
+                        // then try to look for `dev` directory as a "latest" version,
+                        // if we don't find one, we'll report it as a "failure"
```

Comment:
```suggestion
                        // if we don't find one, we'll report it as a "warning"
```

At least that's what the code does.